From ea62f6679db58033b0d3cd4d2f182bf9b0c2bc08 Mon Sep 17 00:00:00 2001
From: J. Bruce Fields <bfields@redhat.com>
Date: Mon, 23 Jun 2014 21:49:51 -0400
Subject: [fs] nfsd: notify_change needs elevated write count

Message-id: <1403560314-32525-50-git-send-email-bfields@redhat.com>
Patchwork-id: 84462
O-Subject: [Patch RHEL7 049/172] nfsd: notify_change needs elevated write count
Bugzilla: 1111719 1115034
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

Looks like this bug has been here since these write counts were
introduced, not sure why it was just noticed now.

Thanks also to Jan Kara for pointing out the problem.

Cc: stable@vger.kernel.org
Reported-by: Matthew Rahtz <mrahtz@rapitasystems.com>
Signed-off-by: J. Bruce Fields <bfields@redhat.com>
Upstream: 9f67f189939eccaa54f3d2c9cf10788abaf2d584
Bugzilla: 1111719
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfsd/vfs.c b/fs/nfsd/vfs.c
index aba35f1..70a949b 100644
--- a/fs/nfsd/vfs.c
+++ b/fs/nfsd/vfs.c
@@ -404,6 +404,7 @@ nfsd_setattr(struct svc_rqst *rqstp, struct svc_fh *fhp, struct iattr *iap,
  umode_t  ftype = 0;
  __be32  err;
  int  host_err;
+ bool  get_write_count;
  int  size_change = 0;
 
  if (iap->ia_valid & (ATTR_ATIME | ATTR_MTIME | ATTR_SIZE))
@@ -411,10 +412,18 @@ nfsd_setattr(struct svc_rqst *rqstp, struct svc_fh *fhp, struct iattr *iap,
  if (iap->ia_valid & ATTR_SIZE)
   ftype = S_IFREG;
 
+ /* Callers that do fh_verify should do the fh_want_write: */
+ get_write_count = !fhp->fh_dentry;
+
  /* Get inode */
  err = fh_verify(rqstp, fhp, ftype, accmode);
  if (err)
   goto out;
+ if (get_write_count) {
+  host_err = fh_want_write(fhp);
+  if (host_err)
+   return nfserrno(host_err);
+ }
 
  dentry = fhp->fh_dentry;
  inode = dentry->d_inode;
-- 
1.7.1